docs/rfcs/rfc-010-cypher-builder.md

```diff
@@ -0,0 +1,186 @@
+* Improve TypeScript types over using plain string.
+* Decoupled from GraphQL and business logic.
+* Static Cypher validation (i.e. before hitting the database).
+* Cypher Backtracking (i.e. allow to easily modify Cypher statements in a non-sequential fashion).
+
+## Proposed interface
+The proposed interface follows a similar [builder-like pattern](https://refactoring.guru/design-patterns/builder) to other query builders. All typings are part of the builder (i.e. none of the library typings are used) to avoid coupling.
+
+The classes will compose the query in a simplified [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) representing our query. Parameters, references and string generation are generated lazily after the full tree has been composed.
+
+## Examples
+
+### Build a simple Match query
+```Cypher
+MATCH(a:Actor)-[:ACTED_IN]->(m:Movie)							 
+WHERE a.name = "Arthur"
+RETURN m.title
+```
+
+Example with plain strings as input
+```typescript
+const actorNode = new CypherBuilder.Node({
+    labels: ["Actor"],
+});
+const movieNode = new CypherBuilder.Node({
+    labels: ["Movie"],
+});
+const relationship = new CypherBuilder.Relationship({ source: node1, target: node2, type: "ACTED_IN" });
+
+
+const query = new CypherBuilder.Match(relationship).where(actorNode, {name: "Arthur"}).return(movieNode, ["title"])
+const { cypher, params } = query.build()
+cypher // "MATCH(this1:Actor)...."
+params // { param1: "Arthur" }
+```
+
+### Query composition
+An important factor to consider is a complex query being composed using multiple statements:
+```cypher
+CALL {
+	CREATE (this0:Movie)
+	SET this0.id = $this0_id
+	RETURN this0
+	}
+RETURN this0 { .id } AS this0
+```
+
+```typescript
+const movieNode = new CypherBuilder.Node({labels: ["Movie"]});
+const idParam = new CypherBuilder.Param();
+
+const subQuery = new CypherBuilder.Query().create(movieNode, {id: idParam}).return(movieNode)
+
+const query = new CypherBuilder.Query().call(subQuery).return(movieNode, {"id"}, "this0")
+
+query.print() // CALL { ...
+query.params // [this0_id]
+```
+
+
+### Backtracking / Lazy
+Because the actual string would be built at the last moment (lazy), all reference can be updated after being defined (e.g. renaming nodes):
+```typescript
+const actorNode = new CypherBuilder.Node({labels: ["Actor"]})
+const movieNode = new CypherBuilder.Node({labels: ["Movie"]})
+const actedIn = new CypherBuilder.Relationship({type: "ACTED_IN", source: actorNode, target: movieNode})
+
+const query = new CypherBuilder.Match(actedIn).where(actorNode, {name: "Arthur"}).return([[movieNode, "title"]])
+
+
+actorNode.addLabel("MyOtherLabel")
+actorNode.setAlias("pepe");
+
+query.print() // MATCH(pepe:Actor:MyOtherLabel)-[:ACTED_IN]->(m:Movie) ....
+
+```
+
+## Risks & Unknowns
+* An [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) for GraphQL to Cypher has yet to be defined. Some aspects of this builder (e.g. parameters) may be depend on the solution agreed for an AST 
```

Comment:
**question (non-blocking)**: Anything in regards to security considerations that needs to be mentioned/added?